npm/pkg/dataplane/ipsets/ipsetmanager_linux.go

```diff
@@ -29,153 +67,409 @@ func (iMgr *IPSetManager) resetIPSets() error {
 		if set.Kind == HashSet {
+			membersToAdd = make(map[string]struct{}, len(set.IPPodKey))
 			for ip := range set.IPPodKey {
-				// TODO add error handler?
-				creator.AddLine(sectionID, nil, util.IpsetAppendFlag, set.HashedName, ip) // add IP
+				membersToAdd[ip] = struct{}{}
 			}
 		} else {
-			setName := setName // to appease golint complaints about function literal
+			membersToAdd = make(map[string]struct{}, len(set.IPPodKey))
 			for _, member := range set.MemberIPSets {
-				memberName := member.Name // to appease golint complaints about function literal
-				errorHandlers := []*ioutil.LineErrorHandler{
-					{
-						Definition: ioutil.NewErrorDefinition(memberSetDoesntExist),
-						Method:     ioutil.SkipLine,
-						Callback: func() {
-							log.Errorf("was going to add member set %s to list %s, but the member doesn't exist", memberName, setName)
-							// TODO handle error
-						},
-					},
-				}
-				creator.AddLine(sectionID, errorHandlers, util.IpsetAppendFlag, set.HashedName, member.HashedName) // add member
+				membersToAdd[member.HashedName] = struct{}{}
+			}
+		}
+
+		// determine which members to add/delete
+		membersToDelete := make(map[string]struct{})
+		for readIndex < len(saveFile) {
+			line, readIndex = parse.Line(readIndex, saveFile)
+			if hasPrefix(line, ipsetCreateString) {
+				break
+			}
+			addMatches := nameForAddRegex.FindSubmatch(line)
+			if len(addMatches) != lengthForSuccessfulMatch || string(addMatches[1]) != hashedName {
+				klog.Errorf("expected an add line for set %s in ipset save file, but got the following line: %s", hashedName, string(line))
+				// TODO send error snapshot
+				line, readIndex = nextCreateLine(readIndex, saveFile)
+				break
 			}
+			restOfLine = line[len(addMatches[0]):]
+			member := string(restOfLine)
+			_, shouldKeep := membersToAdd[member]
+			if shouldKeep {
+				delete(membersToAdd, member) // member already in the kernel, so don't add it later
+			} else {
+				membersToDelete[member] = struct{}{} // member should be deleted from the kernel
+			}
+		}
+
+		// delete undesired members from restore file
+		sectionID := sectionID(addOrUpdateSectionPrefix, prefixedName)
+		for member := range membersToDelete {
+			iMgr.deleteMemberInFile(creator, set, sectionID, member)
+		}
+		// add new members to restore file
+		for member := range membersToAdd {
+			iMgr.addMemberInFile(creator, set, sectionID, member)
+		}
+	}
+}
+
+func nextCreateLine(originalReadIndex int, saveFile []byte) (createLine []byte, nextReadIndex int) {
+	nextReadIndex = originalReadIndex
+	for nextReadIndex < len(saveFile) {
+		createLine, nextReadIndex = parse.Line(nextReadIndex, saveFile)
+		createMatches := nameForCreateRegex.FindSubmatch(createLine)
+		if len(createMatches) == lengthForSuccessfulMatch {
+			return
+		}
+	}
+	return
+}
+
+func haveTypeProblem(set *IPSet, restOfCreateLine []byte) bool {
+	// TODO check maxelem for hash sets? CIDR blocks have a different maxelem
```

Comment:
I am not sure about `TODO`, but just context of maxelem for CIDR is the maximum number of ip address for the hashset -> 2^32 -1 -> 4,294,967,295 based on my understanding. This only works for `ipv4` though

Reply:
clarified